Beautify error notifications

melanieretter opened the pull request:
Introduce a close button in the right upper corner of the error notification.  In case an exception occures the exception message is stored and can be displayed. The standard ErrorHandler of Hawkbit is only assigned if there is no other ErrorHandler assigned yet.

Signed-off-by: Melanie Retter <melanie.retter@bosch-si.com>

schabdo commented on hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/HawkbitUI.java:
```diff
@@ -180,7 +180,10 @@ public void afterViewChange(final ViewChangeEvent event) {
         setLocale(new Locale(locale));
 
-        UI.getCurrent().setErrorHandler(new SPUIErrorHandler());
+        if (UI.getCurrent().getErrorHandler() == null) {
```
Is there a case were you won't set the handler, because it is already set?

melanieretter replied:
Yes. It is necessary if you have a different UI class with a specific errorHandler. The HawkbitUI always overwrites the errorHandler then.

schabdo replied:
I see. Just asking because old code don't do that

schabdo commented on hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/components/HawkbitUIErrorHandler.java:
```diff
@@ -37,29 +32,55 @@
+        notification = new HawkbitNotificationMessage();
         // Build error style
-        final StringBuilder style = new StringBuilder(ValoTheme.NOTIFICATION_FAILURE);
+        style = new StringBuilder(ValoTheme.NOTIFICATION_FAILURE);
```
You can use the fluent API of `StringBuilder` to concat all calls

schabdo commented on hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/components/HawkbitUIErrorHandler.java:
```diff
@@ -37,29 +32,55 @@
+        return exceptionName;
     }
+
+    protected Boolean getAndCheckErrorOrigin(final ErrorEvent event) {
```
Method name is confusing

schabdo commented on hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/components/SPUIErrorHandler.java:
```diff
@@ -39,27 +34,56 @@
+
+    protected Boolean getAndCheckErrorOrigin(final ErrorEvent event) {
+        errorOrigin = findAbstractComponent(event);
+        if (errorOrigin != null && errorOrigin.getUI() != null) {
```
Better: `return (errorOrigin != null && errorOrigin.getUI() != null)`

schabdo commented on hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/components/HawkbitUIErrorHandler.java:
```diff
@@ -0,0 +1,101 @@
+     * Comment for <code>serialVersionUID</code>.
+     */
+    private static final long serialVersionUID = 1877326479308824191L;
+    /**
```
Remove odd comment

schabdo commented on hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/components/HawkbitUIErrorHandler.java:
```diff
@@ -0,0 +1,101 @@
+public class HawkbitUIErrorHandler extends DefaultErrorHandler {
+
+    /**
+     * Comment for <code>serialVersionUID</code>.
```
Remove odd comment

schabdo commented on hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/components/HawkbitUIErrorHandler.java:
```diff
@@ -0,0 +1,101 @@
+            exceptionMessage = error.getMessage();
+            LOG.error("Error in SP-UI:", error);
+        }
+        return exceptionName;
```
You do not return the name from `ErrorEvent` every time! So why I have to provide it for the method call?

schabdo commented on hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/components/HawkbitUIErrorHandler.java:
```diff
@@ -0,0 +1,101 @@
+     *            ErrorEvent
+     * @return boolean if error Origin is not null
+     */
+    protected Boolean getAndCheckErrorOrigin(final ErrorEvent event) {
```
Name is confusing. Can you find a better one?

michahirsch commented:
What's the different between the `HawkbitNotificationMessage` class and the `NotificationMessage` seems like there are doing the same things, right? The `decorate` method is exactly the same.

As I understand the `HawkbitNotificationMessage` is only used for showing error notifications. So either we can declare one notification only to error like `HawkbitErrorNotificationMessage`.

But anyway I don't like the way that the only have a `decorate` method. Why the business logic like showing the notification has been removed from the `HawkbitNotificationMessage` ? 

I think it makes sense to instantiate thy type of notification I need, like `new ErrorNotification()` and the class knows it's style etc.

michahirsch commented:
What has been really changed on style to before? The styles used for the error notification has been the same, don't they?

melanieretter commented:
I renamed `HawkbitNotificationMessage `to `HawkbitErrorNotificationMessage`. The `HakwbitErrorNotificationMessage `is used in the standard errorHandler of the UI. It is used when an unexpected exception occures. The `NotificationMessage `is used when there is an expected (validation) error or notification like saving successful. I removed the decorate method in `HawkbitErrorNotificationMessage `and expand the constructor. The properties style and autoclose has to be in the constructor so that it is possible to overwrite them in case there is a different look desired.

michahirsch commented on hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/components/HawkbitErrorNotificationMessage.java:
```diff
@@ -0,0 +1,43 @@
+    /**
+     * Constructor.
+     */
+    public HawkbitErrorNotificationMessage(final String style, final String caption, final String description,
```
Missing javadoc here.

michahirsch commented on hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/components/HawkbitErrorNotificationMessage.java:
```diff
@@ -0,0 +1,43 @@
+ */
+public class HawkbitErrorNotificationMessage extends Notification {
+
+    /**
```
This javadoc can be removed.
